### DISCUSSION TITLE
How to approach large numbers of groups when RAM is not-huge

### DISCUSSION THREAD

**Discussion by markpayneatwork** [duckdb/duckdb#5746] [Q&A] [answered]
Hi,

My current duckdb v0.6.0 database is around 3.3GB - 580 million rows, made up of 20 million groups of 30 values each. I am trying to process this on a modern laptop, with around 16GB of RAM, but keep crashing the client due to a lack of RAM - the operations are just basic things such as mean by group, etc. What can I do to  help this situation? The table is currently unindexed for example - might that help?

Cheers,

Mark

**Comment by Mytherin** [comment, marked as answer]
Currently DuckDB does not support `GROUP BY` operations where the size of the hash table required to fit all of the groups exceeds RAM - we intend to add support for this in a future release. Note that the hash table stores data *uncompressed*, as opposed to the storage layer that stores data *compressed*. I am guessing that this is the limitation that you are hitting, as 580 million rows at even 32 bytes per row the data would exceed your available memory.

Perhaps you could work around this limitation by dividing the grouping queries into separate queries that each compute a subset of the groups.

**Reply by Mytherin** [reply]
As a side note - if you are working on MacOS (which has support for compressed memory) - it might also be helpful to set the memory limit past your actual physical memory size, as DuckDB computes the memory limit as per the *uncompressed* RAM, whereas the hash table might actually fit in compressed RAM.

e.g. you might be able to get this to work by just disabling the memory limit altogether:

```sql
SET memory_limit='-1';
```

**Reply by markpayneatwork** [reply]
Thanks for the tip - that makes a lot of sense. Sounds like I need to think a bit about chunking the analysis in this case!

Any ideas how far down the road a solution might be?

**Reply by Mytherin** [reply]
I would say it's at least a few months down the road.

**Reply by markpayneatwork** [reply]
Great to know. Thanks!

**Comment by l1t1** [comment]
Does the the data type  effect the size of hash table?
eg.
create table t(a int, b bigint, c int);
when group by a ,does the size different from that of group by b?

**Reply by Mytherin** [reply]
Yes, hash tables are stored uncompressed in memory.

**Comment by l1t1** [comment]
3.3GB - 580 million rows means about 6 bytes per row, are there many empty value?
